<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Molecule Analysis</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="multiomic.html">Multi-omic Integration</a>
        </li>
        <li>
          <a href="nicheclustering.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li>
          <a href="moleculeanalysis.html">Molecule Analysis</a>
        </li>
        <li>
          <a href="pixelanalysis.html">Pixels (Image Only) Analysis</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="interactive.html">Interactive Utilities</a>
        </li>
        <li>
          <a href="importingdata.html">Importing Spatial Data</a>
        </li>
        <li>
          <a href="voltronobjects.html">Working with VoltRon Objects</a>
        </li>
        <li>
          <a href="conversion.html">Converting VoltRon Objects</a>
        </li>
        <li>
          <a href="ondisk.html">OnDisk-based Analysis Utilities</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSB Bioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Molecule Analysis</h1>

</div>


<style>
.title{
  display: none;
}
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
table, th, td {
  border-collapse: collapse;
  align-self: center;
  padding-right: 10px;
  padding-left: 10px;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="molecule-analysis" class="section level1">
<h1>Molecule Analysis</h1>
<p>VoltRon is an end-to-end spatial omics analysis package which also
supports investigating spatial points in sub-cellular resolution,
including those that are based on Fluorescence in situ hybridization
where one can spatially locate each individual molecule that are
hybridized</p>
<p>In this tutorial, we provide a number of examples where we show how
VoltRon manages, analyzes and visualizes molecules as spatial entities
whose identities are given in the metadata of a VoltRon object.</p>
<p><br></p>
<div id="xenium-breast-cancer-data" class="section level2">
<h2>Xenium Breast Cancer Data</h2>
<p>In this use case, we analyse readouts of the experiments conducted on
example tissue sections analysed by the <a
href="https://www.10xgenomics.com/platforms/xenium">Xenium In Situ</a>
platform. Two tissue sections of 5 <span
class="math inline">\(\mu\)</span>m thickness are derived from a single
formalin-fixed, paraffin-embedded (FFPE) breast cancer tissue block.
More information on the spatial datasets and the study can be also be
found on the <a
href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1">BioArxiv
preprint</a>.</p>
<p>You can import these readouts from the <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">10x
Genomics website</a> (specifically, import <strong>In Situ Replicate
1/2</strong>). Alternatively, you can <strong>download a zipped
collection of Xenium readouts</strong> from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/SpatialDataAlignment/Xenium_vs_Visium/10X_Xenium_Visium.zip">here</a>.</p>
<p><br></p>
<div id="creating-voltron-objects" class="section level3">
<h3>Creating VoltRon objects</h3>
<p>VoltRon includes built-in functions for converting readouts of Xenium
experiments into VoltRon objects. The <strong>importXenium</strong>
function uses files of the readout of a Xenium experiment under the
output folder, and forms a VoltRon object.</p>
<p>In this case, the Xenium data includes two assays (in the same layer)
where one is the cell segmentation based cell profile data and the other
is the pure molecule assay where the identity (gene of origin) of RNA
molecules are given. In order to import the molecules, we use
<strong>import_molecules = TRUE</strong>.</p>
<pre class="r watch-out"><code># dependencies
if(!requireNamespace(&quot;RBioFormats&quot;))
  BiocManager::install(&quot;RBioFormats&quot;)
if(!requireNamespace(&quot;arrow&quot;))
  install.packages(&quot;arrow&quot;)
if(!requireNamespace(&quot;rhdf5&quot;))
  BiocManager::install(&quot;rhdf5&quot;)

# import Xenium data
library(VoltRon)
Xen_R1 &lt;- importXenium(&quot;Xenium_R1/outs&quot;, sample_name = &quot;XeniumR1&quot;, import_molecules = TRUE)
Xen_R1</code></pre>
<pre><code>VoltRon Object 
XeniumR1: 
  Layers: Section1 
Assays: Xenium(Main) Xenium_mol </code></pre>
<p><br></p>
</div>
<div id="ondisk-support-optional" class="section level3">
<h3>Ondisk Support <strong>(Optional)</strong></h3>
<p>In order to save space from large amount of data that is generated in
a VoltRon object, you can also save the imported VoltRon object to disk
and still accomplish analysis on molecule datasets. We use the
<strong>saveVoltRon</strong> function which stores both the cell and the
molecule assay on disk.</p>
<pre class="r watch-out"><code>Xen_R1 &lt;- saveVoltRon(Xen_R1, format = &quot;HDF5VoltRon&quot;, 
                      output = &quot;Xen_R1/&quot;, replace = TRUE)</code></pre>
<p>The disk-backed data can be loaded once saved on disk again</p>
<pre class="r watch-out"><code>Xen_R1 &lt;- loadVoltRon(&quot;Xen_R1/&quot;)</code></pre>
<p><br></p>
</div>
<div id="spatial-visualization" class="section level3">
<h3>Spatial Visualization</h3>
<p>With <strong>vrSpatialPlot</strong>, we can visualize Xenium
experiments in subcellular context. We first visualize mRNAs of ACTA2, a
marker for smooth muscle cell actin, as well as markers that are found
in myoepithelial and breast cancer associated tumor cells.</p>
<p>VoltRon incorporates automated rasterization of point (aggregating
points into tiles of equal size) if large number of spatial entities are
being visualized, and thus speeds up plotting.</p>
<pre class="r watch-out"><code>vrSpatialPlot(Xen_R1, group.by = &quot;gene&quot;, assay = &quot;Xenium_mol&quot;, 
              group.ids = c(&quot;ACTA2&quot;, &quot;KRT15&quot;, &quot;TACSTD2&quot;, &quot;CEACAM6&quot;))</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_transcripts_visualize_whole.png" class="center"></p>
<p>We can interactively select a subset of interest within the tissue
section and visualize the localization of these transcripts. Here we
subset a ductal carcinoma niche, and visualize mRNAs of four genes.</p>
<pre class="r watch-out"><code>Xen_R1_subset &lt;- subset(Xen_R1, interactive = TRUE)
Xen_R1_subset &lt;- Xen_R1_subset$subsets[[1]]
vrSpatialPlot(Xen_R1_subset, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;,
              group.ids = c(&quot;ACTA2&quot;, &quot;KRT15&quot;, &quot;TACSTD2&quot;, &quot;CEACAM6&quot;), pt.size = 0.2, legend.pt.size = 5)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_transcripts_visualize.png" class="center"></p>
<p><br></p>
<p>We can use the <strong>n.tile</strong> argument to divide the count
of selected group.ids (e.g. genes here) into hexagonal bins. We will use
this argument to divide the x and y-axis into 200 tiles and aggregate
transcript counts from select <strong>group.id</strong></p>
<pre class="r watch-out"><code>vrSpatialPlot(Xen_R1_subset, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;,
              group.ids = c(&quot;ACTA2&quot;, &quot;KRT15&quot;, &quot;TACSTD2&quot;, &quot;CEACAM6&quot;), n.tile = 200)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/cellspot_transcripts_visualize_bins.png" class="center"></p>
<p><br></p>
</div>
</div>
<div id="xenium-covid-19-lung-data" class="section level2">
<h2>Xenium COVID-19 Lung Data</h2>
<p>Since VoltRon is capable of analyzing localization of sub-cellular
entities such as molecules beyond mRNAs, the analysis may also extend to
extracellular molecules such as viral RNA. Using Xenium in situ, we
locate the open reading frames (ORFs) of genomics and subgenomics RNA of
SARS-COV-2, namely <strong>S2_N</strong> and <strong>S2_orf1ab</strong>
and analyze their abundances over the COVID-19 infected lung tissue.</p>
<p>Specifically, we analyse readouts of <strong>eight tissue micro array
(TMA) tissue sections</strong> that were fitted into the scan area of a
slide loaded on a Xenium in situ instrument. These sections were cut
from <strong>control and COVID-19 lung tissues</strong> of donors
categorized based on disease durations (acute and prolonged). You can
download the standard Xenium output folder <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/Xenium_SARSCOV2.zip">here</a>.</p>
<p>For more information on the TMA blocks, see <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190732">GSE190732</a>
for more information.</p>
<p>Across these eight TMA section, we investigate a section of acute
case which is originated from a lung with extreme number of detected
open reading frames of virus molecules. For convenience, we load a
VoltRon object where cells are already annotated. You can also find the
RDS file <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/acutecase1_annotated.rds">here</a>.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- readRDS(file = &quot;acutecase1_annotated.rds&quot;)</code></pre>
<p>You would see that we have two assays associated with the tissue
section, one for the cells and the other for the molecules.</p>
<pre class="r watch-out"><code>SampleMetadata(vr2_merged_acute1)</code></pre>
<pre><code>            Assay    Layer       Sample
Assay7     Xenium Section1 acute case 1
Assay8 Xenium_mol Section1 acute case 1</code></pre>
<p><br></p>
<div id="spatial-visualization-1" class="section level3">
<h3>Spatial Visualization</h3>
<p>We can visualize individual <strong>S2_N</strong> and
<strong>S2_orf1ab</strong> by simply switching to
<strong>Xenium_mol</strong> assay given</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;, 
              group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), n.tile = 500)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_covid_xenium.png" class="center"></p>
<p><br></p>
<p>You can also visualize both the cellular and subcellular spatial
points on the same plot using <strong>addSpatialLayer</strong> function,
which would show points of the section where virus is abundant with no
cell segment on top.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;, 
              group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), n.tile = 500) |&gt; 
  addSpatialLayer(vr2_merged_acute1, assay = &quot;Xenium&quot;, group.by = &quot;CellType&quot;, plot.segments = TRUE)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/moleculecell_covid_xenium.png" class="center"></p>
<p>Let us zoom into a specific niche to observe the localization of cell
free viral RNA and the surrounding cells.</p>
<pre class="r watch-out"><code>Xen_R1_subset &lt;- subset(Xen_R1, interactive = TRUE)
Xen_R1_subset &lt;- Xen_R1_subset$subsets[[1]]
vrSpatialPlot(Xen_R1_subset, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;, 
              group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), pt.size = 0.4) |&gt; 
  addSpatialLayer(Xen_R1_subset, assay = &quot;Xenium&quot;, group.by = &quot;CellType&quot;, 
                  plot.segments = TRUE, alpha = 0.7)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/moleculecell_covid_xenium_subset.png" class="center"></p>
<p><br></p>
</div>
<div id="hot-spot-analysis" class="section level3">
<h3>Hot Spot Analysis</h3>
<p>Viral RNAs in a highly infected lung sample such as this one may
cluster and aggregate in certain niches of the tissue. We can perform
<strong>hot spot analysis</strong> to find these locations of clusters
with high abundance (i.e. hot spot) to later identify neighboring cells,
and to better understand the effects of the viral infection.</p>
<p>We use the <strong>getHotSpotAnalysis</strong> function to calculate
<a
href="https://en.wikipedia.org/wiki/Getis%E2%80%93Ord_statistics">Getis-Ord</a>
statistics for spatial entities, which detects neighborhoods with large
values of a feature of interest compared to the global mean of the same
feature. Again we are only interested in niches abundant with S2_N and
S2_orf1ab molecules.</p>
<p>In case of molecules, we divide the spatial data into tiles of equal
size and treat the abundance of molecules as a feature. The neighbors of
the tiles are thus other adjacent tiles (eight tiles surrounding each
tile). The tiling parameter <strong>n.tile</strong> can later be changed
by the users.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- getHotSpotAnalysis(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
                                        features = &quot;gene&quot;, 
                                        group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), n.tile = 100)</code></pre>
<p>The <strong>getHotSpotAnalysis</strong> function will produce test
statistics and adjusted p-values for all the selected molecules in the
region.</p>
<pre class="r watch-out"><code>meta &lt;- Metadata(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;)
head(meta[meta$gene %in% c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;),])</code></pre>
<div>
<pre><code style="font-size: 11.7px;">                       id assay_id overlaps_nucleus   gene z_location    qv fov_name nucleus_distance
                   <char>   <char>            <int> <char>      <num> <num>   <char>            <num>
1: 281694020045508_25831b   Assay8                0   S2_N   20.37133    40       Q7         402.4913
2: 281694020045511_25831b   Assay8                0   S2_N   20.39390    40       Q7         362.9937
3: 281694020042882_25831b   Assay8                0   S2_N   20.14611    40       Q7         233.1110
4: 281694020045509_25831b   Assay8                0   S2_N   20.14540    40       Q7         241.3636
5: 281694020042883_25831b   Assay8                0   S2_N   19.92565    40       Q7         170.5991
6: 281694020042884_25831b   Assay8                0   S2_N   19.96041    40       Q7         167.2583
   postfix      Assay    Layer       Sample gene_hotspot_stat   gene_hotspot_padj gene_hotspot_flag
    <char>     <char>   <char>       <char>             <num>               <num>            <char>
1: _25831b Xenium_mol Section1 acute case 1      0.000000e+00                   1              cold
2: _25831b Xenium_mol Section1 acute case 1      0.000000e+00                   1              cold
3: _25831b Xenium_mol Section1 acute case 1      5.752399e-06                   1              cold
4: _25831b Xenium_mol Section1 acute case 1      5.752399e-06                   1              cold
5: _25831b Xenium_mol Section1 acute case 1      1.150483e-05                   1              cold
6: _25831b Xenium_mol Section1 acute case 1      1.150483e-05                   1              cold</code></pre>
</div>
<p><br></p>
<p>Now we can visualize the localization of S2_N and S2_orf1ab molecules
that are located in “hot regions”, or niches that are significantly
abundant with these molecules.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
              group.by = &quot;gene_hotspot_flag&quot;, group.ids = c(&quot;cold&quot;, &quot;hot&quot;), 
              alpha = 1, colors = list(cold = &quot;blue&quot;, hot = &quot;red&quot;), 
              background.color = &quot;white&quot;)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize_virus_hotspot.png" class="center"></p>
<p>The Getis-Ord test statistics and the associated adjusted p-values
are stored within the molecule metadata and can be visualized at
will.</p>
<pre class="r watch-out"><code>library(patchwork)
g1 &lt;- vrSpatialFeaturePlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
                     features = &quot;gene_hotspot_stat&quot;, 
                     alpha = 1, 
                     background.color = &quot;black&quot;)
g2 &lt;- vrSpatialFeaturePlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
                     features = &quot;gene_hotspot_padj&quot;, 
                     alpha = 1, 
                     background.color = &quot;black&quot;)
g1 | g2</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize_virus_hotspot_stats.png" class="center"></p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
