<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Multi-omics</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="multiomic.html">Multi-omic Integration</a>
        </li>
        <li>
          <a href="nicheclustering.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li>
          <a href="moleculeanalysis.html">Molecule Analysis</a>
        </li>
        <li>
          <a href="pixelanalysis.html">Pixels (Image Only) Analysis</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="interactive.html">Interactive Utilities</a>
        </li>
        <li>
          <a href="importingdata.html">Importing Spatial Data</a>
        </li>
        <li>
          <a href="voltronobjects.html">Working with VoltRon Objects</a>
        </li>
        <li>
          <a href="conversion.html">Converting VoltRon Objects</a>
        </li>
        <li>
          <a href="ondisk.html">OnDisk-based Analysis Utilities</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSB Bioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Multi-omics</h1>

</div>


<style>
.title{
  display: none;
}
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
table, th, td {
  border-collapse: collapse;
  align-self: center;
  padding-right: 10px;
  padding-left: 10px;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="multi-omics-analysis" class="section level1">
<h1>Multi-omics Analysis</h1>
<p>VoltRon is an end-to-end spatial multi-omics platform that enables
both automated and manual data alignment workflows to synchronize
coordinate frameworks across spatial omics assays. However, upon
harmonization of the spatial observations, VoltRon can be used to
further transfer and integrate information across omics datasets
captured on same and adjacent tissue sections. These assays are not just
limited to datasets with single cell and spot resolutions but also to
assays solely composed of molecule and image tile observations.</p>
<p>Below, we demonstrate data integration capabilities of VoltRon across
distinct modalities using several use cases. In the <a
href="registration.html">Spatial Data Alignment</a>, we have highlighted
several scenarios to align spatial assays within tissue blocks. However
in this tutorial, we demonstrate analysis starting with data alignment
and further downstream data integration.</p>
<p><br></p>
<div id="xenium-covid-19-lung-data" class="section level2">
<h2>Xenium COVID-19 Lung Data</h2>
<p>Xenium in situ platform is compatible with H&amp;E and
immunofluorescence imaging , thus researchers can stain whole slides
post-Xenium to capture additional information on the morphology of the
tissue and integrate it with the Xenium readouts. In this use case, we
would like to integrate the post-Xenium H&amp;E image of tissue micro
array sections composed of control and COVID-19 lung biopsy punches. Our
objective is to assess the localization of SARS‑CoV‑2 virus molecules by
integrating pathology image annotations.</p>
<p>VoltRon is capable of analyzing molecule-level subcellular datasets
independent of single cells, and specifically those that may also found
outside of these cells. The Xenium run was performed with custom Xenium
probes designed to hybridize distinct open reading frames of SARS‑CoV‑2
virus molecules. These subcellular entities which then can be detected
both within and outside of cells which allows to understand
proliferation mechanics of the virus across the tissue. By aligning the
image and the associated pathology annotations of the post-Xenium
H&amp;E of this slide, we can transfer labels of annotations to Xenium
cells and molecules, thus inferring the abundance of viral RNA
further.</p>
<p>We analyse readouts of <strong>eight tissue micro array (TMA) tissue
sections</strong> that were fitted into the scan area of a slide loaded
on a Xenium in situ instrument. These sections were cut from
<strong>control and COVID-19 lung tissues</strong> of donors categorized
based on disease durations (acute and prolonged). You can download the
standard Xenium output folder <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/Xenium_SARSCOV2.zip">here</a>.
For more information on the TMA blocks, see <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190732">GSE190732</a>
for more information.</p>
<p><br></p>
<div id="single-cells-and-molecules" class="section level3">
<h3>Single Cells and Molecules</h3>
<p>Across these eight TMA section, we investigate a section of acute
case which is originated from a lung with extreme number of detected
open reading frames of virus molecules. For convenience, we load a
VoltRon object where cells are already annotated. You can also find the
RDS file <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/acutecase1_annotated.rds">here</a>.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- readRDS(file = &quot;acutecase1_annotated.rds&quot;)</code></pre>
<p>Lets visualize both the UMAP and Spatial plot of the annotated
cells.</p>
<pre class="r watch-out"><code>vrEmbeddingPlot(vr2_merged_acute1, assay = &quot;Xenium&quot;, embedding = &quot;umap&quot;, 
                group.by = &quot;CellType&quot;, label = TRUE)
vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium&quot;, group.by = &quot;CellType&quot;,
              plot.segments = TRUE)</code></pre>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:45%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_vrembeddingplot.png" class="center">
</td>
<td style="width:55%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize_celltype.png" class="center">
</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>We incorporate two open reading frames (ORFs), named
<strong>S2_N</strong> and <strong>S2_orf1ab</strong>, which represent
unreplicated and replicated virus molecules, respectively. We can
visualize again tile the count of these virus expressions by
specifically selecting these two ORFs.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;, 
              group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), n.tile = 500)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize_virus.png" class="center"></p>
<p><br></p>
<p>We can even zoom into specific locations at the tissue to investigate
virus particles more closely by dropping the <strong>n.tile</strong>
argument and calling interactive visualization.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;, 
              group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), interactive = TRUE, pt.size = 0.1)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecules_visualize_virus_zoom.png" class="center"></p>
<p><br></p>
<p>In the following sections, we will integrate pathological images and
annotations with Xenium datasets to further understand the spatial
localization of virus ORF types over the tissue.</p>
<p><br></p>
</div>
<div id="hot-spot-analysis" class="section level3">
<h3>Hot Spot Analysis</h3>
<p>VoltRon platform allows users to find hot spots (Regions of increased
feature of interest) of several types of spatial entities, for spots,
cells, and even molecules. Please see <a
href="moleculeanalysis.html">Molecule Analysis Tutorial</a> for more
information on hot spot analysis. We switch to the molecule assay of the
VoltRon object, and select virus ORFs to detect regions of high viral
activity.</p>
<pre class="r watch-out"><code>vr2_merged_acute1</code></pre>
<pre><code>VoltRon Object 
acute case 1: 
  Layers: Section1 
Assays: Xenium(Main) Xenium_mol </code></pre>
<p>In this case, we are only interested in niches abundant with S2_N and
S2_orf1ab molecules. We use the <strong>getHotSpotAnalysis</strong>
function to find regions of high S2_N and S2_orf1ab abundance. The
tiling parameter <strong>n.tile</strong> is necessary to apply Getis-Ord
statistics on molecule data since these spatial entities does not have
features in a VoltRon object.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- getHotSpotAnalysis(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
                                        features = &quot;gene&quot;, 
                                        group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), n.tile = 100)</code></pre>
<p>Now we can visualize the localization of molecules that are located
in “hot regions”, or niches that are significantly abundant with these
molecules.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
              group.by = &quot;gene_hotspot_flag&quot;, group.ids = c(&quot;cold&quot;, &quot;hot&quot;), 
              alpha = 1, colors = list(cold = &quot;blue&quot;, hot = &quot;red&quot;), 
              background.color = &quot;white&quot;)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize_virus_hotspot.png" class="center"></p>
</div>
<div id="automated-he-registration" class="section level3">
<h3>Automated H&amp;E Registration</h3>
<p>VoltRon can analyze and also integrate information from distinct
spatial data types such as images, annotations (as regions of interests,
i.e. ROIs) and molecules independently. Using such advanced utilities,
we can make use of histological information and generate new metadata
level information for molecule datasets.</p>
<p>We will first import both histological images and manual annotations
using the <strong>importImageData</strong> function which accepts both
images to generate tile/pixel level datasets but also allows one to
import either a list of segments or <a
href="https://geojson.org/">GeoJSON</a> objects for create ROI-level
datasets as separate assays in a single VoltRon layer. The .geojson file
was generated using <a href="https://qupath.github.io/">QuPath</a> on
the same section H&amp;E image of one Xenium section with the acute
COVID-19 case. We also have to flip the coordinates of ROI annotations
also for they were directly imported from QuPath which incorporates a
reverse coordinate system on the y-axis.</p>
<p>Once imported, the resulting VoltRon object will have two assays in a
single layer, one for tile dataset of the H&amp;E image and the other
for ROI based annotations of again the same image. You can download the
H&amp;E image from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/acutecase1_HE.jpg">here</a>,
and download the json file from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/acutecase1_membrane.geojson">here</a>.</p>
<pre class="r watch-out"><code># get image
imgdata &lt;- importImageData(&quot;acutecase1_HE.jpg&quot;, 
                           segments = &quot;acutecase1_membrane.geojson&quot;, 
                           sample_name = &quot;acute case 1 (HE)&quot;)
imgdata &lt;- flipCoordinates(imgdata, assay = &quot;ROIAnnotation&quot;)
imgdata</code></pre>
<pre><code>VoltRon Object 
acute case 1 (HE): 
  Layers: Section1 
Assays: ImageData(Main) ROIAnnotation 
Features: main(Main) </code></pre>
<p>By visualizing these annotations interactively, we can see that the
ROIs point to the hyaline membranes. Here, we likely find extra-cellular
SARS‑CoV‑2 molecules mostly outside of any single cell.</p>
<pre class="r watch-out"><code>vrSpatialPlot(imgdata, assay = &quot;ROIAnnotation&quot;, group.by = &quot;Sample&quot;, 
              alpha = 0.7, interactive = TRUE)</code></pre>
<p><img width="60%" height="60%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_Covid_HE_zoom.png" class="center"></p>
<p><br></p>
<p>At a first glance, although the Xenium (DAPI) and H&amp;E images are
associated with the same TMA core, they were captured in a different
perspective; that is, one image is almost the 90 degree rotated version
of the other. We will account for this rotation when we (automatically)
align the Xenium data with the H&amp;E image.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- modulateImage(vr2_merged_acute1, brightness = 300, channel = &quot;DAPI&quot;)
vrImages(vr2_merged_acute1, assay = &quot;Assay7&quot;, scale.perc = 20)
vrImages(imgdata, assay = &quot;Assay1&quot;, scale.perc = 20)</code></pre>
<p><img width="90%" height="90%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_twoimages.png" class="center"></p>
<p><br></p>
<p>We now register the H&amp;E image and annotations to the DAPI image
of Xenium section using the <strong>registerSpatialData</strong>
function. We select FLANN (with “Homography” method) automated
registration mode, negate the DAPI image of the Xenium slide, turn 90
degrees to the left and set the scale parameter of both images to
<strong>width = 1859</strong>. See <a href="registration.html">Spatial
Data Alignment</a> tutorial for more information.</p>
<pre class="r watch-out"><code>xen_reg &lt;- registerSpatialData(object_list = list(vr2_merged_acute1, imgdata))</code></pre>
<p><img width="90%" height="90%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_HE_registration.png" class="center"></p>
<p><br></p>
<p>Once registered, we can isolate the registered H&amp;E data and use
it further analysis. We can transfer images and their channels across
assays of one or multiple VoltRon objects. Here, we save the registered
image of the H&amp;E data as an additional channel of Xenium section of
the acuse case 1 sample with molecule data.</p>
<pre class="r watch-out"><code>imgdata_reg &lt;- xen_reg$registered_spat[[2]]
vrImages(vr2_merged_acute1[[&quot;Assay7&quot;]], name = &quot;main&quot;, channel = &quot;H&amp;E&quot;) &lt;- 
  vrImages(imgdata_reg, assay = &quot;Assay1&quot;, name = &quot;main_reg&quot;)
vrImages(vr2_merged_acute1[[&quot;Assay8&quot;]], name = &quot;main&quot;, channel = &quot;H&amp;E&quot;) &lt;- 
  vrImages(imgdata_reg, assay = &quot;Assay1&quot;, name = &quot;main_reg&quot;)</code></pre>
<p>We can now observe the new channels available for the both molecule
and cell-level assays of Xenium data.</p>
<pre class="r watch-out"><code>vrImageChannelNames(vr2_merged_acute1)</code></pre>
<pre><code>               Assay    Layer       Sample  Spatial Channels
Assay7        Xenium Section1 acute case 1     main DAPI,H&amp;E
Assay8    Xenium_mol Section1 acute case 1     main DAPI,H&amp;E</code></pre>
<p><br></p>
<p>You can also add the VoltRon object of H&amp;E data as an additional
assay of the Xenium section such that one layer includes cell,
molecules, ROI Annotations and images in the same time. Specifically, we
add the ROI annotation to the Xenium VoltRon object using the
<strong>addAssay</strong> function where we choose the destination
sample/block and the layer of the assay.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- addAssay(vr2_merged_acute1,
                              assay = imgdata_reg[[&quot;Assay2&quot;]],
                              metadata = Metadata(imgdata_reg, assay = &quot;ROIAnnotation&quot;),
                              assay_name = &quot;ROIAnnotation&quot;,
                              sample = &quot;acute case 1&quot;, layer = &quot;Section1&quot;)
vr2_merged_acute1</code></pre>
<pre><code>VoltRon Object 
acute case 1: 
  Layers: Section1 
Assays: ROIAnnotation(Main) Xenium_mol Xenium </code></pre>
<p><br></p>
</div>
<div id="interactive-visualization" class="section level3">
<h3>Interactive Visualization</h3>
<p>Once the H&amp;E image is registered and transfered to the Xenium
data, we can convert the VoltRon object into an Anndata object (h5ad
file) and use <a href="https://tissuumaps.github.io/">TissUUmaps</a>
tool for interactive visualization.</p>
<pre class="r watch-out"><code># convert VoltRon object to h5ad
as.AnnData(vr2_merged_acute1, assay = &quot;Xenium&quot;, file = &quot;vr2_merged_acute1.h5ad&quot;, 
           flip_coordinates = TRUE, name = &quot;main&quot;, channel = &quot;H&amp;E&quot;)</code></pre>
<p>To run TissUUmaps please follow installation instructions <a
href="https://tissuumaps.github.io/installation/">here</a>, then you can
simply drag and drop both the h5ad file and png file to the
application.</p>
<p><img width="90%" height="90%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_interactivevisualization.png" class="center"></p>
<p><br></p>
</div>
<div id="label-transfer" class="section level3">
<h3>Label Transfer</h3>
<p>Now we can transfer ROI annotations as additional metadata features
of the molecule assay. We will refer the new metadata column as “Region”
which will indicate if the molecule is within any annotated ROI in the
same layer.</p>
<pre class="r watch-out"><code>vrMainAssay(vr2_merged_acute1) &lt;- &quot;ROIAnnotation&quot;
vr2_merged_acute1$Region &lt;- vrSpatialPoints(vr2_merged_acute1)

# set the spatial coordinate system of ROI Annotations assay
vrMainSpatial(vr2_merged_acute1[[&quot;Assay9&quot;]]) &lt;- &quot;main_reg&quot;

# transfer ROI annotations to molecules
vr2_merged_acute1 &lt;- transferData(object = vr2_merged_acute1, from = &quot;Assay9&quot;, to = &quot;Assay8&quot;, 
                                  features = &quot;Region&quot;)

# Metadata of molecules
Metadata(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;)</code></pre>
<div>
<pre><code style="font-size: 11px;">                            id assay_id overlaps_nucleus   gene       qv      Assay    Layer       Sample    Region
                        &lt;char&gt;   &lt;char&gt;            &lt;int&gt; &lt;char&gt;    &lt;num&gt;     &lt;char&gt;   &lt;char&gt;       &lt;char&gt;    &lt;char&gt;
     1: 281651070371256_cb791e   Assay8                0   ENAH 40.00000 Xenium_mol Section1 acute case 1 undefined
     2: 281651070371258_cb791e   Assay8                1  CD274 40.00000 Xenium_mol Section1 acute case 1 undefined
     3: 281651070372515_cb791e   Assay8                0  CD163 40.00000 Xenium_mol Section1 acute case 1 undefined
     4: 281651070374059_cb791e   Assay8                1   CTSL 33.97290 Xenium_mol Section1 acute case 1 undefined
     5: 281651070374411_cb791e   Assay8                0    C1S 40.00000 Xenium_mol Section1 acute case 1 undefined
    ---                                                                                                            
785787: 281874408669584_cb791e   Assay8                0  SFRP2 40.00000 Xenium_mol Section1 acute case 1 undefined
785788: 281874408671410_cb791e   Assay8                0   S2_N 40.00000 Xenium_mol Section1 acute case 1 undefined
785789: 281874408672438_cb791e   Assay8                0 S100A8 40.00000 Xenium_mol Section1 acute case 1 undefined
785790: 281874408673446_cb791e   Assay8                0  TIMP1 29.86642 Xenium_mol Section1 acute case 1 undefined
785791: 281874408673882_cb791e   Assay8                0   S2_N 40.00000 Xenium_mol Section1 acute case 1 undefined</code></pre>
</div>
<p><br></p>
<p>These annotations can be accessed from the default molecule level
metadata of the VoltRon object.</p>
<pre class="r watch-out"><code>vrMainAssay(vr2_merged_acute1) &lt;- &quot;Xenium_mol&quot;
head(table(vr2_merged_acute1$Region))</code></pre>
<pre><code>  ROI1_Assay9  ROI10_Assay9 ROI100_Assay9 ROI101_Assay9 ROI102_Assay9 ROI103_Assay9 
          583           624           784           357           215           200 </code></pre>
<p><br></p>
<p>Now we will grab these annotations from molecule metadata and
calculate the ratio of N to orf1ab frequency of SARS‑CoV‑2 particles
across all annotated molecules.</p>
<pre class="r watch-out"><code>library(dplyr)
s2_summary_hyaline &lt;- 
  Metadata(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;) %&gt;%
  filter(gene %in% c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), 
         Region != &quot;undefined&quot;) %&gt;% 
  summarise(S2_N = sum(gene == &quot;S2_N&quot;), 
            S2_orf1ab = sum(gene == &quot;S2_orf1ab&quot;), 
            ratio = sum(gene == &quot;S2_N&quot;)/sum(gene == &quot;S2_orf1ab&quot;)) %&gt;% 
  as.matrix()
s2_summary_hyaline</code></pre>
<pre><code>      S2_N S2_orf1ab    ratio
[1,] 50977     33532 1.520249</code></pre>
<p><br></p>
</div>
<div id="manual-annotation" class="section level3">
<h3>Manual annotation</h3>
<p>To compare the proportion of <strong>S2_N</strong> and
<strong>S2_orf1ab</strong> molecules in hyaline membranes with tissue
sites of possible infection. We focus on another tissue niche where a
large accumulation of cells with high <strong>S2_N</strong> and
<strong>S2_orf1ab</strong> counts.</p>
<p>By visualizing and zooming on a spatial plot with annotated cells, we
can detect a site of cells with high infection which are also
accompanied by a group of T cells.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium&quot;, group.by = &quot;CellType&quot;, 
              group.ids = c(&quot;H.I. Cells&quot;, &quot;T cells&quot;), plot.segments = TRUE, 
              alpha = 0.6, spatial = &quot;main&quot;, channel = &quot;H&amp;E&quot;, interactive = TRUE)</code></pre>
<p><img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_spatialplot_infected.png" class="center"></p>
<p><br></p>
<p>Now we use the <strong>annotateSpatialData</strong> function to
create an additional annotation of Xenium molecules directly. By
selecting this region of infection we can directly annotate the
‘Xenium_mol’ assay and the metadata of molecules. We use the H&amp;E
image as the background again, and generate a new molecule-level
metadata column called “Infected”.</p>
<pre class="r watch-out"><code>vr2_merged_acute1 &lt;- annotateSpatialData(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, 
                                         label = &quot;Region&quot;, use.image = TRUE, 
                                         channel = &quot;H&amp;E&quot;, annotation_assay = &quot;ROIAnnotation1&quot;)</code></pre>
<p><img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_annotation_infected.png" class="center"></p>
<p><br></p>
</div>
<div id="visualize-annotations" class="section level3">
<h3>Visualize annotations</h3>
<p>Before visualizing the annotations and molecule localizations in the
same time, lets make some changes to the metadata. We basically wanna
change the annotation of all ROIs originated from the GeoJson file to
have the label <strong>Hyaline Membrane</strong>.</p>
<pre class="r watch-out"><code># update molecule metadata
vrMainAssay(vr2_merged_acute1) &lt;- &quot;Xenium_mol&quot;
vr2_merged_acute1$Region &lt;- gsub(&quot;ROI[0-9]+&quot;, &quot;Hyaline Membrane&quot;,
                                 vr2_merged_acute1$Region)

# update ROI metadata
vrMainAssay(vr2_merged_acute1) &lt;- &quot;ROIAnnotation&quot;
vr2_merged_acute1$Region &lt;- gsub(&quot;ROI[0-9]+&quot;, &quot;Hyaline Membrane&quot;, 
                                 vr2_merged_acute1$Region)</code></pre>
<p>Now we can visualize two assays together. We use the
<strong>addSpatialLayer</strong> function to overlay molecule locations
with both the Hyaline Membrane and the infected region annotations.</p>
<pre class="r watch-out"><code>vrSpatialPlot(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;, group.by = &quot;gene&quot;, 
              group.ids = c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), n.tile = 500) |&gt;
  addSpatialLayer(vr2_merged_acute1, assay = &quot;ROIAnnotation&quot;, 
                  group.by = &quot;Region&quot;, alpha = 0.3, 
                  colors = list(`Hyaline Membrane` = &quot;blue&quot;, `Infected` = &quot;yellow&quot;))</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/molecule_visualize_virus_overlay.png" class="center"></p>
</div>
<div id="comparison-of-annotations" class="section level3">
<h3>Comparison of annotations</h3>
<p>By using the newly annotated infection-associated virus molecules, we
can do a comparison of infected regions versus the hyaline
membranes.</p>
<pre class="r watch-out"><code>library(dplyr)
s2_summary_infected &lt;- 
  Metadata(vr2_merged_acute1, assay = &quot;Xenium_mol&quot;) %&gt;%
  filter(gene %in% c(&quot;S2_N&quot;, &quot;S2_orf1ab&quot;), 
         Region == &quot;Infected&quot;) %&gt;% 
  summarise(S2_N = sum(gene == &quot;S2_N&quot;), 
            S2_orf1ab = sum(gene == &quot;S2_orf1ab&quot;), 
            ratio = sum(gene == &quot;S2_N&quot;)/sum(gene == &quot;S2_orf1ab&quot;)) %&gt;% 
  as.matrix()
s2_summary_infected</code></pre>
<pre><code>     S2_N S2_orf1ab    ratio
[1,] 6376      2372 2.688027</code></pre>
<p>Comparison of both the ratio between the infected region and the
hyaline membranes show a considerable difference of the population of
<strong>S2_N</strong> and <strong>S2_orf1ab</strong> molecules.</p>
<pre class="r watch-out"><code>S2_table &lt;- matrix(c(s2_summary_hyaline[,1:2], 
                     s2_summary_infected[,1:2]), 
                   dimnames = list(Region = c(&quot;Hyaline&quot;, &quot;Infected&quot;), 
                                   S2 = c(&quot;N&quot;, &quot;orf1ab&quot;)),
                   ncol = 2,  byrow = TRUE)
S2_table</code></pre>
<pre><code>          S2
Region     N     orf1ab
  Hyaline  50977 33532 
  Infected 6376  2372</code></pre>
<p>A quick test of independence on this contingency table show a
significant difference of ratios across Hyaline and Infected
regions.</p>
<pre class="r watch-out"><code>fisher.test(S2_table, alternative = &quot;two.sided&quot;)</code></pre>
<pre><code>
    Fisher&#39;s Exact Test for Count Data

data:  S2_table
p-value &lt; 2.2e-16
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
 0.5382305 0.5941683
sample estimates:
odds ratio 
 0.5655696 </code></pre>
</div>
</div>
<div id="xenium-breast-cancer-data" class="section level2">
<h2>Xenium Breast Cancer Data</h2>
<p>VoltRon allows analyzing biological images, such as H&amp;E stainings
of tissue sections, independant to omics assays on either the same or
adjacent tissue section. VoltRon defines image tiles, which are
collections of image pixels of the same size, and generate latent
variables that are analyzed downstream. Unsupervised clustering of these
tiles partition regions over the image with respect to morphological
features. Here, these regions might be associated to tumor, immune or
stroma niches can then be transfered to cells in the same tissue section
learn collections of cells with similar features.</p>
<p>We will use the <strong>H&amp;E images</strong> generated from the
same sections as the <strong>Xenium In Situ</strong> platform, and
cluster tiles of the H&amp;E to later transfer groups of tiles to Xenium
cells.</p>
<p>You can download the Xenium readout and the H&amp;E image of the same
tissue section from the <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">10x
Genomics website</a> (specifically, import <strong>In Situ Replicate
1</strong> and <strong>Supplemental: Post-Xenium H&amp;E image
(TIFF)</strong>).</p>
<div id="import-xenium-data" class="section level3">
<h3>Import Xenium Data</h3>
<p>We import both Xenium and post-Xenium H&amp;E images into two
separate VoltRon objects and overlay H&amp;E images.</p>
<pre class="r watch-out"><code># dependencies
if(!requireNamespace(&quot;RBioFormats&quot;))
  BiocManager::install(&quot;RBioFormats&quot;)
if(!requireNamespace(&quot;arrow&quot;))
  install.packages(&quot;arrow&quot;)
if(!requireNamespace(&quot;rhdf5&quot;))
  BiocManager::install(&quot;rhdf5&quot;)

# import Xenium
Xen_R1 &lt;- importXenium(&quot;Xenium_R1/outs&quot;,
                       sample_name = &quot;XeniumR1&quot;, 
                       resolution_level = 3,
                       overwrite_resolution = TRUE)
Xen_R1</code></pre>
<pre class="r watch-out"><code>Xen_R1_image &lt;- importImageData(&quot;Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif&quot;,
                                sample_name = &quot;XeniumR1&quot;,
                                channel_names = &quot;H&amp;E&quot;, 
                                tile.size = 128)
Xen_R1_image                                </code></pre>
<p><br></p>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<p>Using VoltRon we can also integrate tile-level spatial entities and
image tiles with existing image foundation models such as <a
href="https://github.com/beneroth13/dinov2">DinoV2</a> or <a
href="https://github.com/mahmoodlab/UNI">UNI</a>. We can parse tiles
from VoltRon, and get embedding values from a Foundation model to
conduct downstream clustering using PCA and kmeans clustering. See <a
href="pixelanalysis.html#Image_Foundation_Models">Image tiles/pixels
analysis</a> tutorial for more information. You can also download the
DinoV2 embeddings from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/PixelAnalysis/img_128_processed.rds">here</a>.</p>
<pre class="r watch-out"><code>embed_data &lt;- readRDS(&quot;img_128_processed.rds&quot;)
ind &lt;- sapply(vrSpatialPoints(Xen_R1_image), function(x) strsplit(x, split = &quot;_&quot;)[[1]][1])
embed_data &lt;- embed_data[ind,]
rownames(embed_data) &lt;- vrSpatialPoints(Xen_R1_image)
vrEmbeddings(Xen_R1_image, type = &quot;dinov2&quot;, overwrite = TRUE) &lt;-  embed_data</code></pre>
<pre class="r watch-out"><code>Xen_R1_image &lt;- getPCA(Xen_R1_image, dims = 50, data.type = &quot;dinov2&quot;)
Xen_R1_image &lt;- getClusters(Xen_R1_image, method = &quot;kmeans&quot;, label = &quot;Cluster_kmeans&quot;, data.type = &quot;pca&quot;, nclus = 10)</code></pre>
<pre class="r watch-out"><code>vrSpatialPlot(Xen_R1_image, group.by = &quot;Cluster_kmeans&quot;, alpha = 1)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial.png" class="center"></p>
<p><br></p>
</div>
<div id="automated-he-registration-1" class="section level3">
<h3>Automated H&amp;E Registration</h3>
<p>In order to transfer the cluster labels of the tiles that partition
regions, we need to first syncronize the spatial coordinate systems of
the H&amp;E image and the Xenium cells. Here we use the H&amp;E image as
the reference image and warp the coordinate system of the Xenium cells
using the DAPI channel of the Xenium assay. See <a
href="registration.html#Alignment_of_Xenium_and_HE">Spatial Data
Alignment</a> tutorial for more information on image registeration and
alignment.</p>
<pre class="r watch-out"><code>xen_reg &lt;- registerSpatialData(object_list = list(Xen_R1_image, Xen_R1))</code></pre>
<p><img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_HE_registration.png" class="center"></p>
<p><br></p>
<p>We can now parse the registered Xenium data from the result of the
<strong>registerSpatialData</strong> function, and add this assay to the
same section as the H&amp;E image.</p>
<pre class="r watch-out"><code>Xen_R1_reg &lt;- xen_reg$registered_spat[[2]]
Xen_R1_merged &lt;- Xen_R1_image
Xen_R1_merged &lt;- addAssay(Xen_R1_merged, 
                          metadata = Metadata(Xen_R1_reg), 
                          assay = Xen_R1_reg[[&quot;Assay1&quot;]],
                          assay_name = &quot;Xenium&quot;,
                          sample = &quot;XeniumR1&quot;)</code></pre>
</div>
<div id="label-transfer-1" class="section level3">
<h3>Label Transfer</h3>
<p>We can view the combined dataset with the H&amp;E tile data and
Xenium assay that are in the same tissue section.</p>
<pre class="r watch-out"><code>SampleMetadata(Xen_R1_merged)</code></pre>
<pre><code>           Assay    Layer   Sample
Assay1 ImageData Section1 XeniumR1
Assay2    Xenium Section1 XeniumR1</code></pre>
<p>We would like to transfer the labels of the tiles to cells whose
centroids are within a single tile. Thus we transfer
<strong>Cluster_kmeans</strong> from Assay1 (H&amp;E tiles) to Assay2
(Xenium cells). We can also transfer the H&amp;E image to the Xenium
assay the second (registered) coordinate system of Xenium data is sync
with the H&amp;E image.</p>
<pre class="r watch-out"><code># transfer label
Xen_R1_merged &lt;- transferData(Xen_R1_merged, 
                              from = &quot;Assay1&quot;, 
                              to = &quot;Assay2&quot;, 
                              features = &quot;Cluster_kmeans&quot;, 
                              expand = FALSE)

# transfer image
vrImages(Xen_R1_merged[[&quot;Assay2&quot;]], name = &quot;main_reg&quot;, channel = &quot;H&amp;E&quot;) &lt;- 
  vrImages(Xen_R1_merged, assay = &quot;Assay1&quot;)</code></pre>
<p>You can visualize the transferred tile labels in two different
spatial coordinate systems and with different images (DAPI, H&amp;E
etc.) using <strong>spatial</strong> and <strong>channel</strong>
arguments, respectively.</p>
<pre class="r watch-out"><code>vrMainAssay(Xen_R1_merged) &lt;- &quot;Xenium&quot;
vrSpatialPlot(Xen_R1_merged, group.by = &quot;Cluster_kmeans&quot;, spatial = &quot;main&quot;)
vrSpatialPlot(Xen_R1_merged, group.by = &quot;Cluster_kmeans&quot;, spatial = &quot;main_reg&quot;, 
              channel = &quot;H&amp;E&quot;)</code></pre>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:50%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial_cell.png" class="center">
</td>
<td style="width:50%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial_cell_he.png" class="center">
</td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
<div id="marker-analysis" class="section level3">
<h3>Marker Analysis</h3>
<p>Now that we have labelled cells whose partitioning is determined by
the morphological (image foundation model) features, we can perform
marker analysis to determine signatures associated with groups. We
convert the cells assays of the VoltRon object to a Seurat object prior
to marker analysis, find markers associated with groups and then parse
significant markers to visualize later. See <a
href="conversion.html#Seurat">Conversion</a> tutorial for more
information on VoltRon’s interoperability features.</p>
<pre class="r watch-out"><code>library(Seurat)
Xen_data_seu &lt;- VoltRon::as.Seurat(Xen_R1_merged, cell.assay = &quot;Xenium&quot;, type = &quot;image&quot;)
Xen_data_seu &lt;- NormalizeData(Xen_data_seu, scale.factor = 100)
Idents(Xen_data_seu) &lt;- &quot;Cluster_kmeans&quot;
markers_image &lt;- FindAllMarkers(Xen_data_seu)
markers_image_top &lt;- markers_image %&gt;%
  group_by(cluster) %&gt;%
  dplyr::filter(avg_log2FC &gt; 1, p_val_adj &lt; 0.05, pct.1 &gt; 0.5)
head(as.data.frame(markers_image_top))</code></pre>
<div>
<pre><code style="font-size: 13px;">  p_val avg_log2FC pct.1 pct.2 p_val_adj cluster  gene
1     0   1.679807 0.787 0.335         0       8  MLPH
2     0   1.406676 0.869 0.430         0       8  KRT8
3     0   1.442903 0.782 0.359         0       8  CDH1
4     0   1.693801 0.808 0.385         0       8  FASN
5     0   1.526783 0.874 0.458         0       8 EPCAM
6     0   1.493334 0.670 0.259         0       8 MYO5B</code></pre>
</div>
<p><br></p>
<p>Heatmap visualization of these clusters reveal markers of tumor,
stroma and immune niches.</p>
<pre class="r watch-out"><code>Xen_R1_merged &lt;- normalizeData(Xen_R1_merged, sizefactor = 1000)
vrHeatmapPlot(Xen_R1_merged, group.by = &quot;Cluster_kmeans&quot;, 
              features = unique(markers_image_top$gene), 
              show_row_names = TRUE)</code></pre>
<p><img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_heatmap_cell.png" class="center"></p>
<p><br></p>
<p>Inspecting the heatmap and the spatial plot simultanuously, we can
see that some tiles are associated with the H&amp;E background where
others are representing tumor, immune and stroma niches.</p>
<pre class="r watch-out"><code>annotation &lt;- Xen_R1_merged$Cluster_kmeans 
annotation[annotation %in% c(6,8)] &lt;- &quot;Tumor Niche&quot;
annotation[annotation %in% c(1)] &lt;- &quot;Immune Niche&quot;
annotation[annotation %in% c(2,10,9)] &lt;- &quot;Stroma&quot;
annotation[annotation %in% c(4)] &lt;- &quot;Stroma&quot;
annotation[annotation %in% c(3,5,7)] &lt;- &quot;Background&quot;
Xen_R1_merged$annotation &lt;- annotation</code></pre>
<p>Users can also visualize tile and cell simultaneously and their
features.</p>
<pre class="r watch-out"><code># select a subset
Xen_R1_image_subset &lt;- subset(Xen_R1_merged, interactive = TRUE)
Xen_R1_image_subset &lt;- Xen_R1_image_subset$subsets[[1]]

# visualize 
vrSpatialPlot(Xen_R1_image_subset, group.by = &quot;Cluster_kmeans&quot;, alpha = 0.7)
vrSpatialPlot(Xen_R1_image_subset, assay = &quot;Xenium&quot;, group.by = &quot;annotation&quot;, 
              alpha = 0.7, plot.segments = TRUE, channel = &quot;H&amp;E&quot;)</code></pre>
<p><img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial_cell_joint_subset.png" class="center"></p>
<pre class="r watch-out"><code>vrSpatialFeaturePlot(Xen_R1_image_subset, assay = &quot;Xenium&quot;, 
                     features = c(&quot;PTPRC&quot;, &quot;EPCAM&quot;, &quot;MMP2&quot;), 
                     alpha = 0.7, plot.segments = TRUE, ncol = 2)</code></pre>
<p><img width="85%" height="85%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial_cell_markers_subset.png" class="center"></p>
<p><br></p>
</div>
</div>
<div id="xenium-if-tonsil-data" class="section level2">
<h2>Xenium + IF Tonsil Data</h2>
<p>Automated alignment workflow of VoltRon can also be used to integrate
modalities generated from the same tissue section. In this use case, we
will perform multiplex immunofluorescence (IF) imaging after a Xenium
experiment and integrate Protein and RNA modalities for clustering.</p>
<div id="import-xenium-data-1" class="section level3">
<h3>Import Xenium Data</h3>
<p>We will generate Xenium and IF dataset from a human tonsil tissue
section. We first import the Xenium data, and interactively subset the
tonsil sample from the tissue microarray.</p>
<p>You can also download the tonsil subset from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/Human_Tonsil/Xenium_tonsil_subset.rds">here</a>.</p>
<pre class="r watch-out"><code># dependencies
if(!requireNamespace(&quot;RBioFormats&quot;))
  BiocManager::install(&quot;RBioFormats&quot;)
if(!requireNamespace(&quot;arrow&quot;))
  install.packages(&quot;arrow&quot;)
if(!requireNamespace(&quot;rhdf5&quot;))
  BiocManager::install(&quot;rhdf5&quot;)

# import Xenium
xen &lt;- importXenium(&quot;Xenium/outs&quot;, 
                    sample_name = &quot;XeniumTonsil&quot;,
                    resolution_level = 3, 
                    overwrite_resolution = TRUE)

# subset interactive
xen &lt;- subset(xen, interactive = TRUE)
xen_tonsil &lt;- xen$subsets[[1]]

# load from RDS if needed
# xen_tonsil &lt;- readRDS(&quot;Xenium_tonsil_subset.rds&quot;)

# view image
vrImages(xen_tonsil)</code></pre>
<p><img width="60%" height="60%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_xen.png" class="center"></p>
<p><br></p>
</div>
<div id="import-if-images" class="section level3">
<h3>Import IF Images</h3>
<p>Xenium in situ platform allows same section multiplex imaging post
experiment. Depending on the protocol, the intensity measures of each
channel can be stored in an ome.tiff file. We use the
<strong>importImageData</strong> to import the images of channels where
we can select channels and even give them names.</p>
<p>We import only the DAPI, CD20 and CD21 channels. We will use DAPI
channel for aligning these images to Xenium data, and remaining channels
for clustering with the RNA features from Xenium.</p>
<p>You can download the ome.tiff file of the same section IF experiment
<a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/Human_Tonsil/same_section_IF.ome.tif">here</a>.</p>
<pre class="r watch-out"><code>if_tonsil &lt;- importImageData(&quot;same_section_IF.ome.tif&quot;, tile.size = 100, 
                             series = 1, resolution = 1, channels = c(1,4,8), 
                             channel_names = c(&quot;DAPI&quot;, &quot;CD20&quot;, &quot;CD21&quot;), 
                             is.RGB = FALSE)</code></pre>
<p><br></p>
</div>
<div id="automated-alignment" class="section level3">
<h3>Automated Alignment</h3>
<p>Now, we perform automated alignment across Xenium and IF
datasets.</p>
<pre class="r watch-out"><code>xen_tonsil &lt;- modulateImage(xen_tonsil, brightness = 500)
if_tonsil &lt;- modulateImage(if_tonsil, brightness = 700, channel = &quot;DAPI&quot;)
xen_reg &lt;- registerSpatialData(object_list = list(if_tonsil, xen_tonsil))</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_register.png" class="center"></p>
<p><br></p>
<p>The output of the <strong>registerSpatialData</strong> function will
include a VoltRon object of the Xenium data whose cell segments are
syncronized to the IF experiment.</p>
<pre class="r watch-out"><code># get registered data
xen_tonsil &lt;- xen_reg$registered_spat[[2]]</code></pre>
<p>We can also transfer the CD20, and CD21 channels to the registered
coordinate space of the Xenium data.</p>
<pre class="r watch-out"><code>vrImages(xen_tonsil[[&quot;Assay1&quot;]], name = &quot;main_reg&quot;, channel = &quot;CD20&quot;) &lt;- 
  vrImages(if_tonsil, channel = &quot;CD20&quot;)
vrImages(xen_tonsil[[&quot;Assay1&quot;]], name = &quot;main_reg&quot;, channel = &quot;CD21&quot;) &lt;- 
  vrImages(if_tonsil, channel = &quot;CD21&quot;)</code></pre>
<p><br></p>
</div>
<div id="quantify-intensity-of-cells" class="section level3">
<h3>Quantify Intensity of Cells</h3>
<p>With VoltRon, we can export the registered segments, and save it as a
GeoJson file for further processing with QuPath, which we will use it to
quantify the mean intensity measures of CD20 and CD21 protein per Xenium
segments.</p>
<p>This will enable collecting same segment RNA and Protein measures for
multi-modal clustering.</p>
<pre class="r watch-out"><code>segts &lt;- vrSegments(flipCoordinates(xen_tonsil))
generateGeoJSON(segts, &quot;registered_segments.geojson&quot;)</code></pre>
<p>To perform intensity quntification of cells, simply drag and drop the
ome.tiff and geojson files to QuPath. You can then navigate to
<strong>Analyze -&gt; Calculate Features -&gt; Add Intensity
Features</strong>, and click Run after selecting markers where we select
CD20 and CD21. If you wanna perform the mean and median intensity
calculation of features, you can use the GeoJSON file with registered
segments found <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/Human_Tonsil/registered_segments.geojson">here</a>.</p>
<p>After calculating mean and median intensities, we navigate to
<strong>Measure -&gt; Show Annotation Measurements</strong> and click
<strong>Save</strong> to write the measurements to a .txt file.</p>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_QuPath.png" class="center"></p>
<p><br></p>
<p>After quantifying with QuPath, we can import these measures and save
it as a new feature data in VoltRon object. You can find the IF
measurements of CD20 and CD21 markers <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Multiomics/Human_Tonsil/IF_measurements.txt">here</a>.</p>
<pre class="r watch-out"><code># add feature from QuPath
datax &lt;- read.table(&quot;IF_measurements.txt&quot;, 
                    header = TRUE, sep = &quot;\t&quot;)
rownames(datax) &lt;- datax$Name
datax &lt;- datax[vrSpatialPoints(xen_tonsil),]
datax &lt;- datax[,c(11,13)]
datax &lt;- apply(datax, 2, function(x){
  x[is.na(x)] &lt;- min(x, na.rm = TRUE)
  x
})

# adjust channel intensities
datax[,1] &lt;- ifelse(datax[,1] &lt; 1054, datax[,1],1054)
datax[,1] &lt;- ifelse(datax[,1] &gt; 331, datax[,1],331)
datax[,1] &lt;- datax[,1] - min(datax[,1])
datax[,2] &lt;- ifelse(datax[,2] &lt; 1421, datax[,2],1421)
datax[,2] &lt;- ifelse(datax[,2] &gt; 491, datax[,2],491)
datax[,2] &lt;- datax[,2] - min(datax[,2])
colnames(datax) &lt;- c(&quot;CD20&quot;, &quot;CD21&quot;)</code></pre>
<p>We use the <strong>addFeature</strong> function to import the new
Protein measures as additional feature sets</p>
<pre class="r watch-out"><code># add as protein features
xen_tonsil &lt;- addFeature(xen_tonsil, data = t(datax), feature_name = &quot;Protein&quot;)
xen_tonsil</code></pre>
<pre><code>VoltRon Object 
XeniumTonsil: 
  Layers: Section1 
Assays: Xenium(Main) 
Features: RNA(Main) Protein </code></pre>
<p><br></p>
</div>
<div id="processing-modalities" class="section level3">
<h3>Processing Modalities</h3>
<p>We first do a simple filtering on the Xenium cells, remove all cells
whose total mRNA counts are lower than 30, and we normalize these counts
afterwards.</p>
<pre class="r watch-out"><code># subset
spatialpoints &lt;- vrSpatialPoints(xen_tonsil)[as.vector(Metadata(xen_tonsil)$Count &gt; 30)]
xen_tonsil &lt;- subset(xen_tonsil, spatialpoints = spatialpoints)

# normalize RNA
xen_tonsil &lt;- normalizeData(xen_tonsil, sizefactor = 1000)</code></pre>
<p>VoltRon also includes methods to process protein intensity measures.
We use the hyperbolic arcsine transformation with scaling parameter
0.2.</p>
<pre class="r watch-out"><code># normalize Protein markers
vrMainFeatureType(xen_tonsil) &lt;- &quot;Protein&quot;
xen_tonsil &lt;- normalizeData(xen_tonsil, method = &quot;hyper.arcsine&quot;, scale = 0.2)</code></pre>
<p>We can visualize spatial localization of both normalized Protein and
RNA features. CD20 and MS4A1 are Protein and RNA synonyms of the same
marker, hence they largely co-localize. On the other hand, CD38 and CD21
are both markers of germinal centers of human tonsils, but CD21 mainly
expressed in these centers and CD38 can also be expressed in plasma
cells</p>
<pre class="r watch-out"><code>library(patchwork)
vrMainFeatureType(xen_tonsil) &lt;- &quot;Protein&quot;
g1 &lt;- vrSpatialFeaturePlot(xen_tonsil, features = c(&quot;CD20&quot;, &quot;CD21&quot;), 
                           background.color = &quot;black&quot;, n.tile = 300, ncol = 2, 
                           norm = TRUE)
vrMainFeatureType(xen_tonsil) &lt;- &quot;RNA&quot;
g2 &lt;- vrSpatialFeaturePlot(xen_tonsil, features = c(&quot;MS4A1&quot;, &quot;CD38&quot;), 
                           background.color = &quot;black&quot;, n.tile = 300, ncol = 2)
g1 / g2</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_spatial.png" class="center"></p>
<p><br></p>
</div>
<div id="ondisk-optional" class="section level3">
<h3>OnDisk (Optional)</h3>
<p>VoltRon provides additional utilities to save object on disk to lower
memory usage of data analysis and clustering. We can save the object on
disk and continue analysis which will potentially increase the
clustering speed.</p>
<pre class="r watch-out"><code>xen_tonsil &lt;- saveVoltRon(xen_tonsil, 
                               format = &quot;HDF5VoltRon&quot;, 
                               output = &quot;data/OnDisk/Xenium_IF&quot;, 
                               replace = TRUE)</code></pre>
<p><br></p>
</div>
<div id="clustering-1" class="section level3">
<h3>Clustering</h3>
<p>We can now apply dimensional reduction which will allow us to cluster
cells. The <strong>feat_type</strong> argument in the
<strong>getPCA</strong> function will accept multiple modality names of
a VoltRon object. Here, <strong>vrFeatureTypeNames</strong> will provide
the names of all modalities and will result in <strong>getPCA</strong>
computing PC dimensions using both RNA and Protein modalities.</p>
<pre class="r watch-out"><code>xen_tonsil &lt;- getPCA(xen_tonsil, 
                     feat_type = vrFeatureTypeNames(xen_tonsil),
                     dims = 30)
xen_tonsil &lt;- getUMAP(xen_tonsil, dims = 1:30)</code></pre>
<p>Let us visualize the expression of RNA and Protein features. We can
see that CD21 (Protein) and CD38 (RNA) features, which both are germinal
center markers, are expressed in different combinations. Also, CD20 and
its RNA counterpart MS4A1 show similar expression patterns.</p>
<pre class="r watch-out"><code># visualize embeddings 
vrMainFeatureType(xen_tonsil) &lt;- &quot;Protein&quot;
g1 &lt;- vrEmbeddingFeaturePlot(xen_tonsil, 
                       features = c(&quot;CD20&quot;, &quot;CD21&quot;), embedding = &quot;umap&quot;, 
                       pt.size = 0.4)
vrMainFeatureType(xen_tonsil) &lt;- &quot;RNA&quot;
g2 &lt;- vrEmbeddingFeaturePlot(xen_tonsil, 
                       features = c(&quot;MS4A1&quot;, &quot;CXCL13&quot;), embedding = &quot;umap&quot;, 
                       pt.size = 0.4)
g1 / g2</code></pre>
<p><img width="100%" height="100%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_embedding.png" class="center"></p>
<p><br></p>
<p>Clustering with multi-modal PCA will reveal clusters high in both RNA
and Protein features. We first cluster the cells using SNN based graph
clustering, and then visualize the expression of CD21 to detect marker
specific clusters.</p>
<pre class="r watch-out"><code># clustering
xen_tonsil &lt;- getProfileNeighbors(xen_tonsil, 
                                  dims = 1:30, method = &quot;SNN&quot;)
xen_tonsil &lt;- getClusters(xen_tonsil, resolution = 1.0, 
                          label = &quot;Clusters&quot;, graph = &quot;SNN&quot;) 

# visualize
vrEmbeddingPlot(xen_tonsil, group.by = &quot;Clusters&quot;, embedding = &quot;umap&quot;, 
                pt.size = 0.4, label = TRUE)</code></pre>
<p><img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_embedding_cluster.png" class="center"></p>
<p><br></p>
<p>As observed, there is one cluster with high CD21 protein
expression.</p>
<pre class="r watch-out"><code>vrViolinPlot(xen_tonsil, features = &quot;CD21&quot;, group.by = &quot;Clusters&quot;)</code></pre>
<p><img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_violin_cluster.png" class="center"></p>
<p><br></p>
<p>We can visualize the spatial localization of this cluster alone,
which point to the germinal centers of the human tonsil.</p>
<pre class="r watch-out"><code>vrSpatialPlot(xen_tonsil, group.by = &quot;Clusters&quot;, group.ids = 14, 
              plot.segments = TRUE)</code></pre>
<p><img width="80%" height="80%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/multiomic_XeniumIF_spatial_gc.png" class="center"></p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
