<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Importing Spatial Data</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">VoltRon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="tutorials.html">Explore</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignette
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Spatial Data Integration</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="registration.html">Spatial Data Alignment</a>
        </li>
        <li>
          <a href="multiomic.html">Multi-omic Integration</a>
        </li>
        <li>
          <a href="nicheclustering.html">Niche Clustering</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Downstream Analysis</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="roianalysis.html">ROI Analysis</a>
        </li>
        <li>
          <a href="spotanalysis.html">Cell/Spot Analysis</a>
        </li>
        <li>
          <a href="moleculeanalysis.html">Molecule Analysis</a>
        </li>
        <li>
          <a href="pixelanalysis.html">Pixels (Image Only) Analysis</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Utilities</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="interactive.html">Interactive Utilities</a>
        </li>
        <li>
          <a href="importingdata.html">Importing Spatial Data</a>
        </li>
        <li>
          <a href="voltronobjects.html">Working with VoltRon Objects</a>
        </li>
        <li>
          <a href="conversion.html">Converting VoltRon Objects</a>
        </li>
        <li>
          <a href="ondisk.html">OnDisk-based Analysis Utilities</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-envelope-o"></span>
     
    Contact
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://bioinformatics.mdc-berlin.de">Altuna Lab/BIMSB Bioinfo</a>
    </li>
    <li>
      <a href="https://www.mdc-berlin.de/landthaler">Landthaler Lab/BIMSB</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-github"></span>
     
    GitHub
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="https://github.com/BIMSBbioinfo/VoltRon">VoltRon</a>
    </li>
    <li>
      <a href="https://github.com/BIMSBbioinfo">BIMSB Bioinfo</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Importing Spatial Data</h1>

</div>


<style>
.title{
  display: none;
}
body {
  text-align: justify
}
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
table, th, td {
  border-collapse: collapse;
  align-self: center;
  padding-right: 10px;
  padding-left: 10px;
}
</style>
<style type="text/css">
.watch-out {
  color: black;
}
</style>
<p><br></p>
<div id="analysis-of-images" class="section level1">
<h1>Analysis of Images</h1>
<p>VoltRon is a spatial omics analysis platform that allows storing a
large amount of spatially resolved datasets. As opposed to datasets with
either supracellular (spot-level) or cellular resolutions, one might
analyze image datasets and pixels to characterize the morphology of a
tissue section.</p>
<p>Here, images can also be used to build VoltRon objects where pixels
(or tiles) are defined as spatial points, and then can also be used for
multiple downstream analysis purposes.</p>
<p><br></p>
<div id="importing-images-he" class="section level2">
<h2>Importing Images (H&amp;E)</h2>
<p>In this use case, we will analyze the H&amp;E image derived from a
tissue section that was first analyzed by The 10x Genomics <a
href="https://www.10xgenomics.com/platforms/xenium">Xenium In Situ</a>
platform. Three tissue sections were cut from a single formalin-fixed,
paraffin-embedded (FFPE) breast cancer tissue block. A 5 <span
class="math inline">\(\mu\)</span>m section was used to generate a
single Xenium replicate. The H&amp;E staining was performed on the
tissue after the Xenium run.</p>
<p>More information on the Xenium and the study can be also be found on
the <a
href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1">BioArxiv
preprint</a>. You can download the H&amp;E image from the <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">10x
Genomics website</a> as well (specifically, import the
<strong>Post-Xenium H&amp;E image (TIFF)</strong> data).</p>
<p>We incorporate <strong>importImageData</strong> to convert an image
into a tile/pixel-based spatial dataset.</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- importImageData(&quot;Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif&quot;,
                                sample_name = &quot;XeniumR1image&quot;, 
                                image_name = &quot;H&amp;E&quot;, tile.size = 100)
Xen_R1_image</code></pre>
<pre><code>VoltRon Object 
XeniumR1image: 
  Layers: Section1 
Assays: ImageData(Main) 
Features: main(Main) </code></pre>
<pre class="r watch-out"><code>vrImages(Xen_R1_image, scale.perc = 2)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/importdata_HE.png" class="center"></p>
<p></br></p>
</div>
<div id="building-datasets-for-tiles" class="section level2">
<h2>Building Datasets for Tiles</h2>
<p>The VoltRon object stores the metadata information and localization
of all tiles in the image. These tiles, like any observation, can be
analyzed independently where you can reduce its dimensionality to
perform clustering later.</p>
<pre class="r watch-out"><code>head(Metadata(Xen_R1_image))</code></pre>
<pre><code>             id assay_id     Assay    Layer        Sample
         &lt;char&gt;   &lt;char&gt;    &lt;char&gt;   &lt;char&gt;        &lt;char&gt;
1: tile1_2c05fc   Assay1 ImageData Section1 XeniumR1image
2: tile2_2c05fc   Assay1 ImageData Section1 XeniumR1image
3: tile3_2c05fc   Assay1 ImageData Section1 XeniumR1image
4: tile4_2c05fc   Assay1 ImageData Section1 XeniumR1image
5: tile5_2c05fc   Assay1 ImageData Section1 XeniumR1image
6: tile6_2c05fc   Assay1 ImageData Section1 XeniumR1image</code></pre>
<p>Each row is associated with a tile of parsed from the image, and the
users can define additional metadata columns afterwards.</p>
<pre class="r watch-out"><code>nrow(Metadata(Xen_R1_image))</code></pre>
<pre><code>[1] 24888</code></pre>
<p>In order to generate data values from image tiles, we can treat each
pixel of a tile as a data dimension (in this case, by 200x200 tiles we
have 40000 pixels, thus dimensions). We generate this dataset as
below:</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- generateTileData(Xen_R1_image)
dim(vrData(Xen_R1_image))</code></pre>
<pre><code>[1] 40000 13974</code></pre>
<p></br></p>
</div>
<div id="dimensionality-reduction" class="section level2">
<h2>Dimensionality Reduction</h2>
<p>We can now reduce the dimensionality of these pixel intensities using
PCA and visualize the data variability across tiles. The data matrix we
generated with the <strong>generateTileData</strong> earlier will be
used here similar to gene/protein expression datasets, where we reduce
40000 variables (each associated with a pixel) to a manageable number,
say 15.</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- getPCA(Xen_R1_image, dims = 15)
Xen_R1_image &lt;- getUMAP(Xen_R1_image)
vrEmbeddingPlot(Xen_R1_image, embedding = &quot;umap&quot;, group.by = &quot;Sample&quot;)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_embedding.png" class="center"></p>
<p></br></p>
</div>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Here, the dimensionally reduced image tiles can be clustered using
either graph clustering approaches or kmeans clustering depending on the
choice.</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- getProfileNeighbors(Xen_R1_image, data.type = &quot;pca&quot;, 
                                    k = 10, method = &quot;SNN&quot;, dims = 1:15)
Xen_R1_image &lt;- getClusters(Xen_R1_image, resolution = 1.0, 
                            label = &quot;Cluster_snn&quot;, graph = &quot;SNN&quot;)</code></pre>
<p>Here, we visualize the transparent clustering information of tiles by
rasterizing the tiles into larger sizes to increase plotting speed where
we can see the distinction between different tissue niches of the
image.</p>
<pre class="r watch-out"><code>g1 &lt;- vrSpatialPlot(Xen_R1_image, group.by = &quot;Cluster_snn&quot;, 
                    alpha = 0.5, n.tile = 100)
g2 &lt;- vrEmbeddingPlot(Xen_R1_image, embedding = &quot;umap&quot;, 
                      group.by = &quot;Cluster_snn&quot;)
g1 | g2 </code></pre>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:60%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_spatialcluster.png" class="center">
</td>
<td style="width:40%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_embeddingcluster.png" class="center">
</td>
</tr>
</tbody>
</table>
<p></br></p>
</div>
</div>
<div id="image-foundation-models" class="section level1">
<h1>Image Foundation Models</h1>
<p>Using VoltRon we can also integrate tile-level spatial entities and
image tiles with existing image foundation models such as <a
href="https://github.com/beneroth13/dinov2">DinoV2</a> or <a
href="https://github.com/mahmoodlab/UNI">UNI</a>.</p>
<p>We can parse tiles from VoltRon, and get embedding values from a
Foundation model to conduct downstream clustering using PCA and kmeans
clustering.</p>
<p><br></p>
<div id="importing-images-he-1" class="section level2">
<h2>Importing Images (H&amp;E)</h2>
<p>We first import the H&amp;E image to VoltRon. This time we will set
the tile size to 128 which will generate 34400 tiles on the breast
cancer tissue.</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- importImageData(&quot;Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif&quot;,
                                sample_name = &quot;XeniumR1image&quot;,
                                image_name = &quot;H&amp;E&quot;, tile.size = 128)
Xen_R1_image@metadata</code></pre>
<pre><code>VoltRon Metadata Object 
This object includes: 
   34400 tiles </code></pre>
<p><br></p>
</div>
<div id="add-external-embeddings" class="section level2">
<h2>Add External Embeddings</h2>
<p>We will now load and add the DinoV2 embeddings as additional
embedding data before analysis and clustering. Here, we use
<strong>overwrite</strong> argument (default is <code>FALSE</code>) as a
safe-guard for overwriting already existing embeddings with the same
name, i.e. <code>type = "dinov2"</code>, but if the an embedding with
name <code>dinov2</code> doesn’t exist, this will be overlooked. You can
download the DinoV2 embeddings from <a
href="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/PixelAnalysis/img_128_processed.rds">here</a>.</p>
<pre class="r watch-out"><code># get embed data
dinov2_embed &lt;- readRDS(&quot;img_128_processed.rds&quot;)
ind &lt;- sapply(vrSpatialPoints(Xen_R1_image), function(x) strsplit(x, split = &quot;_&quot;)[[1]][1])
dinov2_embed &lt;- dinov2_embed[ind,]
rownames(dinov2_embed) &lt;- vrSpatialPoints(Xen_R1_image)

# import DinoV2 embedding
vrEmbeddings(Xen_R1_image, type = &quot;dinov2&quot;, overwrite = TRUE) &lt;-  dinov2_embed</code></pre>
<p><br></p>
</div>
<div id="dimensionality-reduction-1" class="section level2">
<h2>Dimensionality Reduction</h2>
<p>If we check the number of extracted features from the DinoV2 model,
we see that its 1536 which we could further reduced using PCA and UMAP
prior to clustering.</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- getPCA(Xen_R1_image, dims = 50, data.type = &quot;dinov2&quot;)
Xen_R1_image &lt;- getUMAP(Xen_R1_image)
vrEmbeddingPlot(Xen_R1_image, embedding = &quot;umap&quot;, group.by = &quot;Sample&quot;)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_embedding.png" class="center"></p>
<p><br></p>
</div>
<div id="clustering-1" class="section level2">
<h2>Clustering</h2>
<p>With the PCA embeddings generated from the DinoV2 embeddings now we
can apply k-means clustering.</p>
<pre class="r watch-out"><code>Xen_R1_image &lt;- getClusters(Xen_R1_image, method = &quot;kmeans&quot;, 
                            label = &quot;Cluster_kmeans&quot;, data.type = &quot;pca&quot;, 
                            nclus = 10)</code></pre>
<pre class="r watch-out"><code>vrEmbeddingPlot(Xen_R1_image, embedding = &quot;umap&quot;, group.by = &quot;Cluster_snn&quot;)
vrSpatialPlot(Xen_R1_image, group.by = &quot;Cluster_kmeans&quot;, alpha = 1)</code></pre>
<table>
<tbody>
<tr style="vertical-align: center">
<td style="width:45%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_embedding_kmeans.png" class="center">
</td>
<td style="width:55%; vertical-align: center">
<img src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial.png" class="center">
</td>
</tr>
</tbody>
</table>
<p>We can zoom to certain locations over the tissue and visualize the
tile-based embedding clusters over regions.</p>
<pre class="r watch-out"><code># interactive subsetting
Xen_R1_image_subset &lt;- subset(Xen_R1_image, interactive = TRUE)
Xen_R1_image_subset &lt;- Xen_R1_image_subset$subsets[[1]]

# visualization
vrSpatialPlot(Xen_R1_image_subset, group.by = &quot;Cluster_kmeans&quot;, alpha = 0.5)</code></pre>
<p><img width="70%" height="70%" src="https://bimsbstatic.mdc-berlin.de/landthaler/VoltRon/Package/images/pixel_dinov2_spatial_subset.png" class="center"></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
